Propagate exceptions when bootstrapping a management server during the AddNode workflow.

PavelZaytsev opened the pull request:
## Overview

Description:

Why should this be merged: 

There is a join performed on the bootstrapManagementServer in the AddNode SendRequest method. If it fails, it throws the exception wrapped in the CompletionException, which does not get handled on the upper layers. Even if it is not wrapped in this exception, the root exception is still wrapped in the RuntimeException, so it still wont get caught. The goal is to propagate TimeOut and Network exceptions from this call, so it can get retried. 

Related issue(s) (if applicable): #<number>


## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java:
```diff
@@ -99,17 +100,26 @@ public void handleFailure(IReconfigurationHandlerPolicy failureHandlerPolicy,
+        catch (TimeoutException toe){
+            future.completeExceptionally(toe);
+            return future;
+        }
```
This seems hacky, and besides why this is needed?  `bootstrapLayoutServer` has an exceptionally block:

```java
.exceptionally(throwable -> {
        try {
                CFUtils.unwrap(throwable, AlreadyBootstrappedException.class);
        } catch (AlreadyBootstrappedException e) {
                log.info("bootstrapLayoutServer: Layout Server {} already bootstrapped.", endpoint);
        }
        return true;
})
```
Notice this `CFUtils.unwrap(throwable, AlreadyBootstrappedException.class);` It will throw a `RutimeException` in case of exceptions other than `AlreadyBootstrappedException`, which makes the returned future exceptionally completed . Thus in this `bootstrapNewNode` method you will get a exceptionally completed future of `RutimeException` with a `TimeoutException` wrapped inside.

pankti-m commented on runtime/src/main/java/org/corfudb/runtime/view/workflows/AddNode.java:
```diff
@@ -46,7 +46,12 @@ public AddNode(@NonNull String endpointToAdd, @NonNull CorfuRuntime runtime,
         Layout layout = new Layout(runtime.getLayoutView().getLayout());
-        runtime.getManagementView().bootstrapManagementServer(nodeForWorkflow, layout).join();
+        boolean bootstrapped =
+                runtime.getManagementView().bootstrapManagementServer(nodeForWorkflow, layout);
```
shouldnt timeoutException thrown from bootstrapManagementServer be caught here?

PavelZaytsev replied:
Outdated

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:
```diff
@@ -54,18 +55,22 @@ public static void bootstrap(@NonNull Layout layout,
+        } catch (TimeoutException toe){
+            log.error("BootstrapUtil: Layout Server {}:{} timeout occurred.",
+                    router.getHost(), router.getPort());
+            throw toe;
```
I don't understand why we need to explicitly handle TimeoutException here. What about NetworkExceptions and other exceptions? Besides, the caller of this, namely BootstrapUtil.bootstrap() already logs and handles all exceptions: https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java#L128-L134

pankti-m commented on test/src/test/java/org/corfudb/runtime/view/ManagementViewTest.java:
```diff
@@ -1640,7 +1640,7 @@ public void testPartialBootstrapNodeSuccess() throws ExecutionException, Interru
         // Attempt bootstrapping the node. The node should attempt bootstrapping both the components Layout Server and
         // Management Server.
-        assertThat(corfuRuntime.getLayoutManagementView().bootstrapNewNode(SERVERS.ENDPOINT_1).get()).isTrue();
+        assertThat(corfuRuntime.getLayoutManagementView().bootstrapNewNode(SERVERS.ENDPOINT_1).get()).isFalse();
```
you can change the test comments to reflect this new behaviour

PavelZaytsev replied:
Outdated

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/view/LayoutView.java:
```diff
@@ -319,21 +319,26 @@ public void committed(long epoch, Layout layout, boolean force)
+            return false;
+        } catch (TimeoutException te) {
+            log.error("bootstrapLayoutServer: Timeout occurred.");
+            throw te;
```
Why specifically handling TimeoutException, there are other exceptions like NetworkException right? As explained in previous comment, the existing bootstrapLayoutServer implementation does return an exceptionally completed future due to CFUtil.unwrap(), I think the problem might not be here.

PavelZaytsev replied:
Ok, I see. I won't touch it then. I'll leave management and layout views as is but will unwrap the exception in the AddNode sendRequest I guess.

pankti-m commented on runtime/src/main/java/org/corfudb/runtime/view/workflows/AddNode.java:
```diff
@@ -46,7 +46,12 @@ public AddNode(@NonNull String endpointToAdd, @NonNull CorfuRuntime runtime,
+        boolean bootstrapped =
+                runtime.getManagementView().bootstrapManagementServer(nodeForWorkflow, layout);
+
+        if (!bootstrapped) {
```
minor - the variable name bootstrapped is confusing here.  If the server is bootstrapped, we log the warning.  But the check appears to be doing the opposite.  You can change the name to something else.

PavelZaytsev replied:
Outdated

annym commented:
@PavelZaytsev Can you please add as description a clear explanation of the bug/issue? perhaps a clear example of what we have seen eases the review and also for reference in the future is a good idea.

PavelZaytsev commented:
@annym I have updated the description.
@WenbinZhu I have addressed the comments and updated the description.

Please review when you get the chance.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/workflows/WorkflowRequest.java:
```diff
@@ -141,7 +141,7 @@ public void invoke() {
                 }
 
-            } catch (NetworkException | TimeoutException | IllegalStateException e) {
+            } catch (Exception e) {
```
I don't think swallowing all exceptions is a good idea. The retry only needs to happen on timeout and network exceptions.

Maithem replied:
Also, if this thread is interrupted, the interrupt flag isn't restored.

PavelZaytsev replied:
Ok, I'll bring it back in.

Maithem commented on runtime/src/main/java/org/corfudb/runtime/view/workflows/AddNode.java:
```diff
@@ -46,7 +47,7 @@ public AddNode(@NonNull String endpointToAdd, @NonNull CorfuRuntime runtime,
         // Bootstrap a management server first.
         Layout layout = new Layout(runtime.getLayoutView().getLayout());
-        runtime.getManagementView().bootstrapManagementServer(nodeForWorkflow, layout).join();
+        CFUtils.getUninterruptibly(runtime.getManagementView().bootstrapManagementServer(nodeForWorkflow, layout));
```
You can pass Network/Timeout exceptions types.

PavelZaytsev replied:
Ok, I see. I'll pass the Network and TimeOut types here and return the 3 exceptions.

PavelZaytsev commented:
@annym 
@Maithem 
@pankti-m 
@WenbinZhu 

I have addressed the comments and updated the description.
Please review when you get the chance.

WenbinZhu commented on runtime/src/main/java/org/corfudb/runtime/view/workflows/AddNode.java:
```diff
@@ -44,9 +46,12 @@ public AddNode(@NonNull String endpointToAdd, @NonNull CorfuRuntime runtime,
 
     @Override
     protected UUID sendRequest(@NonNull ManagementClient managementClient) throws TimeoutException {
```
I think later on there should a cleanup patch on all the inherited children of sendRequest because others are not taking care of NetworkException, and the signature suggests it only throws TimeoutException.

PavelZaytsev replied:
Sounds good.

WenbinZhu commented:
@PavelZaytsev please rebase the commits.
